exec.go

```diff
@@ -126,57 +125,50 @@ func setPrimaryKey(rv reflect.Value, id int64) {
 // the record in the DB with one Exec.
 // The given data needs to be:
 //
-// *[]*strcut
+// *[]*struct
 // *[]struct
 // []*struct
 // []struct
 //
 // sqlpro will executes one INSERT statement per call.
 func (db *DB) InsertBulk(table string, data interface{}) error {
-	var (
-		rv         reflect.Value
-		structMode bool
-		err        error
-	)
-
-	rv, structMode, err = checkData(data)
-	if err != nil {
+	rv, structMode, err := checkData(data)
+	switch {
+	case err != nil:
 		return err
-	}
-
-	if structMode {
+	// i don't know if we really want to fail silently
+	// returning back an error will probably be a good thing
+	// for the user
+	case rv.Len() == 0:
+		return nil
+	case structMode:
 		return fmt.Errorf("InsertBulk: Need Slice to insert bulk.")
 	}
 
-	key_map := make(map[string]*fieldInfo, 0)
-	rows := make([]map[string]interface{}, 0)
-
-	if rv.Len() == 0 {
-		return nil
-	}
+	key_map := make(map[string]*fieldInfo)
+	rows := make([]map[string]interface{}, rv.Len())
 
-	for i := 0; i < rv.Len(); i++ {
+	for i := range rows {
 		row := reflect.Indirect(rv.Index(i)).Interface()
-
 		values, structInfo, err := db.valuesFromStruct(row)
 
 		if err != nil {
 			return xerrors.Errorf("sqlpro.InsertBulk error: %w", err)
 		}
 
-		rows = append(rows, values)
+		rows[i] = values
```

Comment:
See above, I think its better to use 0, cap + append. (not sure though)

Reply:
It is certainly an interesting thing to to discuss.
I have dissasembled both and it's interesting to see the generated instructions both.
https://play.golang.org/p/rzmor_Hmrw2
We can see that the length == 0 with cap is performing a lot worse, as it needs a lot bigger stack, doing comparisons at every append, to check if it needs to grow (it doubles in size everytime it reaches the limit) and finally calls to `CALL runtime.growslice(SB)`
What do you think?

Reply:
I think you are right, may be we should benchmark this? For now I am happy with the indexed version.

Reply:
Yes, i will benchmark and post the results but i think we shouldn't include the benchmark code to this branch because it would be out of the scope of the task.

Reply:
> BenchmarkMakeLen-4         40288             48000 ns/op
> BenchmarkMakeCap-4        110458            114600 ns/op
> PASS